ktor-http/ktor-http-cio/common/src/io/ktor/http/cio/websocket/WebSocketExtension.kt

```diff
@@ -0,0 +1,149 @@
+/*
+ * Copyright 2014-2020 JetBrains s.r.o and contributors. Use of this source code is governed by the Apache 2.0 license.
+ */
+
+package io.ktor.http.cio.websocket
+
+import io.ktor.util.*
+
+/**
+ * WebSocket extensions API is experimental according to [KTOR-688](https://youtrack.jetbrains.com/issue/KTOR-688)
+ * To get more information about Ktor experimental guarantees consult
+ * with [KTOR-1035](https://youtrack.jetbrains.com/issue/KTOR-1035).
+ */
+@RequiresOptIn(
+    message = "WebSocket extensions API is experimental according to []()." +
+        "To get more information about Ktor experimental guarantees consult " +
+        "with [KTOR-1035](https://youtrack.jetbrains.com/issue/KTOR-1035)",
+    level = RequiresOptIn.Level.ERROR
+)
+public annotation class ExperimentalWebSocketExtensionApi
+
+@OptIn(ExperimentalWebSocketExtensionApi::class)
+private typealias ExtensionInstaller = () -> WebSocketExtension<*>
+
+/**
+ * Factory that defines WebSocket extension. The factory is used in pair with [WebSocketExtensionsConfig.install]
+ * method to install WebSocket extension in client or server.
+ *
+ * Usually this interface implemented in `companion object` of the origin [WebSocketExtension].
+ */
+@ExperimentalWebSocketExtensionApi
+public interface WebSocketExtensionFactory<ConfigType: Any, ExtensionType: WebSocketExtension<ConfigType>> {
+    /**
+     * Key is used to locate extension.
+     */
+    public val key: AttributeKey<ExtensionType>
+
+    /**
+     * First extension bit used by current extension.
+     *
+     * This flag is used to detect extension conflicts: only one feature with enabled flag is allowed.
+     * To set the flag value please consult with specification of the extension you're using.
+     */
+    public val rsv1: Boolean
```

Comment:
can we use `false` as the default value here?

Reply:
It's the design question actually. I would like to explicitly ask for setting `rsv` bits

Reply:
Correct me if I am wrong but in practice, people will override 0 or 1 flag in their extensions. That's why I prefer to have a default value `false` here.